klayout/drc/testing/run_regression.py

```diff
@@ -0,0 +1,924 @@
+                    line_list = line.split('"')
+                    rule_info = dict()
+                    rule_info["table_name"] = os.path.basename(runset).replace(".drc", "")
+                    rule_info["rule_name"] = line_list[1]
+                    rule_info["in_rule_deck"] = 1
+                    rules_data.append(rule_info)
+
+    df = pd.DataFrame(rules_data)
+    df.drop_duplicates(inplace=True)
+    df.to_csv(os.path.join(output_path, "rule_deck_rules.csv"), index=False)
+    return df
+
+
+def analyze_test_patterns_coverage(rules_df, tc_df, output_path):
+    """
+    This function analyze the test patterns before running the test cases.
+
+    Parameters
+    ----------
+    rules_df : pd.DataFrame
+        DataFrame that holds all the rules that are found in the rule deck.
+    tc_df : pd.DataFrame
+        DataFrame that holds all the test cases and all the information required.
+    output_path : string or Path
+        Path of the run location to store the output analysis file.
+
+    Returns
+    -------
+    pd.DataFrame
+        A DataFrame with analysis of the rule testing coverage.
+    """
+    cov_df = (
+        tc_df[["table_name", "rule_name"]]
+        .groupby(["table_name", "rule_name"])
+        .count()
+        .reset_index(drop=False)
+    )
+    cov_df = cov_df.merge(rules_df, on="rule_name", how="outer")
+    cov_df["runset"].fillna("", inplace=True)
+    cov_df.to_csv(os.path.join(output_path, "testcases_coverage.csv"), index=False)
+
+    return cov_df
+
+
+def generate_merged_testcase(orignal_testcase, marker_testcase):
+    """
+    This function will merge orignal gds file with generated
+    markers gds file.
+
+    Parameters
+    ----------
+    orignal_testcase : string or Path object
+        Path string to the orignal testcase
+
+    marker_testcase : string or Path
+        Path of the output marker gds file generated from db file.
+
+    Returns
+    -------
+    merged_gds_path : string or Path
+        Path of the final merged gds file generated.
+    """
+
+    new_lib = gdstk.Library()
+
+    lib_org = gdstk.read_gds(orignal_testcase)
+    lib_marker = gdstk.read_gds(marker_testcase)
+
+    # Getting flattened top cells
+    top_cell_org = lib_org.top_level()[0].flatten(apply_repetitions=True)
+    top_cell_marker = lib_marker.top_level()[0].flatten(apply_repetitions=True)
+    marker_polygons = top_cell_marker.get_polygons(apply_repetitions=True, include_paths=True, depth=None)
+
+    # Merging all polygons of markers with original testcase
+    for marker_polygon in marker_polygons:
+        top_cell_org.add(marker_polygon)
+
+    # Adding flattened merged cell
+    new_lib.add(top_cell_org.flatten(apply_repetitions=True))
+
+    # Writing final merged gds file
+    merged_gds_path = f'{marker_testcase.replace(".gds", "")}_merged.gds'
+    new_lib.write_gds(merged_gds_path)
+
+    return merged_gds_path
+
+
+def draw_polygons(polygon_data, cell, lay_num, lay_dt, path_width):
```

Comment:
curious if the same could be achieve with `klayout.ly`?

Reply:
BTW, about speed and memory requirements, we did try klayout and it took too much time to try to regenerate the lyrdb to gds. gdstk is extremely fast.

Reply:
Let's consolidate this discussion with the other comment.